### PR TITLE
[quantization] Introduction of mixed calibration dataset

### PR DESCRIPTION
**Pull request by Torrero**
This PR introduces support for mixed calibration dataset that combines both image-text samples (e.g., VQAv2, COCO) and text-only samples (e.g., Wikitext2, Alpaca). This enables more comprehensive calibration for Qwen3-VL.

Text-only samples requires additional processing in `tico/quantization/algorithm/qwen3_vl_gptq/quantizer.py`

Additionally `--nsamples_for_qcalibration` argument was extended to comma-separated list. 
Formats supported:
- Just a number N -> uses default dataset "vqav2" with N samples
- "dataset:n_samples" -> uses default split
- "dataset:split:n_samples" -> uses specified split
- Multiple datasets: "vqav2:testdev:128,wikitext2:train:128"

Related issue:
- https://github.com/Samsung/TICO/issues/645

---------------------
#### 1. __quantizer.py__ (qwen3_vl_gptq) 

- Added __separate caches for vision batches__: `_vision_cache_args`, `_vision_cache_kwargs`, `_num_vision_batches`
- Vision stages (patch_embed, blocks, merger) now __only process batches with vision inputs__ via `vision_only=True` flag
- Text decoder layers process __all batches__ (both image-text and text-only)
- Prevents errors when text-only inputs lack image tokens
- Added warning when no vision inputs are found in calibration data

#### 2. __utils.py__ (qwen3_vl_gptq) - Cache Handling

- Enhanced `append_batch_to_cache()` to maintain __index alignment__ across batches with different kwargs
- New keys are backfilled with `None` for previous batches
- Missing keys in current batch get `None` placeholder
- Ensures consistent batch indexing across heterogeneous samples

#### 3. __utils.py__ (gptq) - Gradient Safety

- Added check to __skip modules without gradients__ during sensitivity calibration
- Prevents errors when vision modules don't participate in forward pass (text-only inputs)

#### 4. __vlm_eval_utils.py__ - Dataset Support

- Added __text-only datasets__: `alpaca` and adapters

- New `is_text_only` flag in dataset registry to distinguish dataset types

- New functions:

  - `build_text_only_inputs()` - builds processor inputs for text-only data
  - `_build_text_calib_inputs()` - samples random fixed-length sequences from text
  - `get_mixed_calib_inputs()` - combines multiple datasets into single calibration set

#### 5. __quantize_qwen3_vl_with_gptq.py__ 

- Changed `--nsamples_for_qcalibration` from `int` to `str` type

- Supports multiple formats:

  - Just a number: `128` → uses default `vqav2:testdev:128`
  - `dataset:n_samples` → uses default split
  - `dataset:split:n_samples` → uses specified split
  - Multiple: `vqav2:testdev:128,wikitext2:train:128,coco:32`

- Added `load_calib_inputs()` function to parse and load multiple datasets

TICO-DCO-1.0-Signed-off-by:  Evgenii Maltsev <e.maltsev@samsung.com>

### CODE REVIEW

File: tico/quantization/evaluation/vlm_eval_utils.py [hunk 1]
```diff
@@ -856,3 +909,183 @@ def get_calib_inputs(
...
+
+    Returns:
+        A list of processor output objects from all datasets.
+    """
+    calib_inputs = []
+
+    for dataset, config in dataset_config.items():
+        n_samples = config.get("n_samples", None)
+        split = config.get("split", None)
+        if n_samples <= 0:
```
Comment (dvsav) [inline tico/quantization/evaluation/vlm_eval_utils.py]: `n_samples = config.get("n_samples", None)` assumes `n_samples` may become `None`.
`if n_samples <= 0` will then raise a `TypeError`.
How about this:
```python
if n_samples is None or n_samples <= 0:
    continue
```

File: tico/quantization/evaluation/vlm_eval_utils.py [hunk 2]
```diff
@@ -856,3 +909,183 @@ def get_calib_inputs(
...
+                seed=seed,
+            )
+            calib_inputs.extend(text_inputs)
+        else:
+            # For image-text datasets: take first n_samples directly
+            if adapter is None:
+                print(f"[warn] No adapter for dataset '{dataset}', skipping")
+                continue
+
+            count = 0
```
Comment (dvsav) [inline tico/quantization/evaluation/vlm_eval_utils.py]: Unused variable

**Comment by Torrero** [comment]
Benchmarks evaluation for Qwen3-vl-4B with mixed calibrated datasets: 'vqav2:testdev:128,wikitext2:train:128' 

| Config ID| PPL | vqav2 | CIDEr | Bleu_4 |
|--|--|--|--|--|
| FP32                       | 10.54 | 0.895   | 0.361 | 0.025 |
| GPTQ_SMSE_w4A16_smoothQuant_vision | 13.09 | 0.7210  | 0.173 | 0.019 |
| GPTQ_MSE_w4A16_smoothQuant_vision | 12.21 | 0.6980  | 0.169 | 0.018 |

File: tico/quantization/algorithm/qwen3_vl_gptq/utils.py [hunk 3]
```diff
@@ -515,15 +515,36 @@ def append_batch_to_cache(
         cache_kwargs: Keyword cache storage.
         *args: Positional batch values.
         **kwargs: Keyword batch values.
+
```
Comment (mhs4670go) [inline tico/quantization/algorithm/qwen3_vl_gptq/utils.py:527]: I think this cache alignment logic still has a corner case for kwargs-only calibration inputs.

The problem is not that `move_batch_to_device()` returns a dict. That part is fine. The issue is that calibration calls the model as `q_m(**dev_inp)`, so `m_args` is empty and all inputs are stored through `cache_kwargs`. In that case, `cache_args` remains empty, and `expected_batch_idx` becomes `0` even if previous kwargs-only batches have already been cached.

This can break alignment when a text-only batch appears before a vision batch.

Example:

```python
cache_args = []
cache_kwargs = {}

# batch 0: text-only
append_batch_to_cache(
    cache_args,
    cache_kwargs,
    input_ids="text_ids",
    attention_mask="text_mask",
)

# batch 1: image-text
append_batch_to_cache(
    cache_args,
    cache_kwargs,
    input_ids="vlm_ids",
    attention_mask="vlm_mask",
    pixel_values="image_tensor",
    image_grid_thw="grid",
)
```

Expected cache:

```python
cache_kwargs["input_ids"]      == ["text_ids", "vlm_ids"]
cache_kwargs["attention_mask"] == ["text_mask", "vlm_mask"]
cache_kwargs["pixel_values"]   == [None, "image_tensor"]
cache_kwargs["image_grid_thw"] == [None, "grid"]
```

But with the current implementation, because `cache_args` is empty, a newly introduced key is initialized with:

```python
cache_kwargs[key] = [None] * expected_batch_idx  # expected_batch_idx == 0
```

So pixel_values may become:
```python
cache_kwargs["pixel_values"] == ["image_tensor"]
```

This means the logical batch indices are no longer aligned across kwargs. Later, when replaying batch 0, the code may combine text-only `input_ids` with the first vision `pixel_values`, or hit an index error when accessing a later batch.

This is especially relevant for mixed calibration because the user can pass datasets in either order, e.g.:

```
--nsamples_for_qcalibration wikitext2:train:128,vqav2:testdev:128
```

In this order, text-only samples are cached first and vision keys appear later.

A safer fix would be to derive the previous batch count from the existing cache contents, not only from cache_args:

```python
def _num_cached_batches(
    cache_args: list[list[Any]],
    cache_kwargs: dict[str, list[Any]],
) -> int:
    lengths = [len(v) for v in cache_args]
    lengths.extend(len(v) for v in cache_kwargs.values())
    return max(lengths, default=0)

def append_batch_to_cache(
    cache_args: list[list[Any]],
    cache_kwargs: dict[str, list[Any]],
    *args: Any,
    **kwargs: Any,
) -> None:
    prev_batches = _num_cached_batches(cache_args, cache_kwargs)

    for idx, item in enumerate(args):
        if idx >= len(cache_args):
            cache_args.append([None] * prev_batches)
        cache_args[idx].append(detach_clone_tree(item))

    for key in list(cache_kwargs.keys()):
        if key not in kwargs:
            cache_kwargs[key].append(None)

    for key, value in kwargs.items():
        if key not in cache_kwargs:
            cache_kwargs[key] = [None] * prev_batches
        cache_kwargs[key].append(detach_clone_tree(value))
```

File: tico/quantization/evaluation/vlm_eval_utils.py [hunk 4]
```diff
@@ -856,3 +909,183 @@ def get_calib_inputs(
...
+
+        print(
+            f"[info] Loading calibration data from '{dataset}' (split={split}): {n_samples} samples"
+        )
+
+        if dataset not in DATASETS:
+            print(f"[warn] Unknown dataset '{dataset}', skipping")
+            continue
+
+        ds, adapter = get_dataset(dataset=dataset, n=n_samples, split=split)
```
Comment (mhs4670go) [inline tico/quantization/evaluation/vlm_eval_utils.py]: For text-only datasets, `n_samples` is used both as the number of source dataset rows to read and as the number of calibration windows to generate.

For example, `wikitext2:train:128` loads **only 128 source rows**, concatenates them, and then samples 128 windows from that concatenated text.

File: tico/quantization/evaluation/vlm_eval_utils.py [hunk 5]
```diff
@@ -856,3 +909,183 @@ def get_calib_inputs(
...
+                    processor=processor,
+                    image=item["image"],
+                    question=item["question"],
+                    return_tensors="pt",
+                    max_seq_len=max_seq_len,
+                )
+                calib_inputs.append(inputs)
+
+    print(f"[info] Total calibration samples: {len(calib_inputs)}")
+    return calib_inputs
```
Comment (mhs4670go) [inline tico/quantization/evaluation/vlm_eval_utils.py:1103]:
```suggestion
    if not calib_inputs:
        raise ValueError(
            "No calibration inputs were loaded. "
            "Please check --nsamples_for_qcalibration."
        )

    print(f"[info] Total calibration samples: {len(calib_inputs)}")
    return calib_inputs
```

File: tico/quantization/evaluation/vlm_eval_utils.py [hunk 6]
```diff
@@ -856,3 +909,183 @@ def get_calib_inputs(
...
+        n_samples = config.get("n_samples", None)
+        split = config.get("split", None)
+        if n_samples <= 0:
+            continue
+
+        print(
+            f"[info] Loading calibration data from '{dataset}' (split={split}): {n_samples} samples"
+        )
+
+        if dataset not in DATASETS:
```
Comment (mhs4670go) [inline tico/quantization/evaluation/vlm_eval_utils.py:1035]: Seems that `{"dataset:split": n_samples}` key will be unknown. Is it intentional?

Parsing could be done like this instead.

```python
dataset_specs = [
    {"name": "wikitext2", "split": "train", "n_samples": 128},
    {"name": "wikitext2", "split": "test", "n_samples": 128},
]

```

File: tico/quantization/evaluation/vlm_eval_utils.py [hunk 7]
```diff
@@ -856,3 +909,183 @@ def get_calib_inputs(
...
+            print(f"[warn] Unknown dataset '{dataset}', skipping")
+            continue
+
+        ds, adapter = get_dataset(dataset=dataset, n=n_samples, split=split)
+        is_text_only = DATASETS[dataset].get("is_text_only", False)
+
+        if is_text_only:
```
Comment (mhs4670go) [inline tico/quantization/evaluation/vlm_eval_utils.py:1041]: Text only inputs are just concatenated unlike vision inputs that are chagned with chat templates. I think we need to change text inputs, too. But, seems the code change would be big. Let's make it TODO.

File: tico/quantization/evaluation/vlm_eval_utils.py (hunk 6 above)
Comment (Torrero) [reply tico/quantization/evaluation/vlm_eval_utils.py:1035]: Ahh, I've just forgot to change the comments, thank you. Actually input format is following: 
```
{"dataset": {"split":"split","n_samples": n_samples}}
```

File: tico/quantization/evaluation/vlm_eval_utils.py (hunk 4 above)
Comment (Torrero) [reply tico/quantization/evaluation/vlm_eval_utils.py]: fixed it

File: tico/quantization/evaluation/vlm_eval_utils.py (hunk 7 above)
Comment (Torrero) [reply tico/quantization/evaluation/vlm_eval_utils.py:1041]: Added `TODO`

File: tico/quantization/evaluation/vlm_eval_utils.py (hunk 2 above)
Comment (Torrero) [reply tico/quantization/evaluation/vlm_eval_utils.py]: fixed it

File: tico/quantization/evaluation/vlm_eval_utils.py (hunk 5 above)
Comment (Torrero) [reply tico/quantization/evaluation/vlm_eval_utils.py:1103]: Changed

File: tico/quantization/evaluation/vlm_eval_utils.py (hunk 1 above)
Comment (Torrero) [reply tico/quantization/evaluation/vlm_eval_utils.py]: fixed it

File: tico/quantization/algorithm/qwen3_vl_gptq/utils.py (hunk 3 above)
Comment (Torrero) [reply tico/quantization/algorithm/qwen3_vl_gptq/utils.py:527]: Thank you, I fixed it

**Review by mhs4670go — APPROVED** [review]
LGTM

**Review by dvsav — APPROVED** [review]
LGTM
